### PR TITLE
Add scheduler predictor

### LINKED ISSUE #62: Ability to predict if more pods can be scheduled to the cluster
We need a tool to predict if new pods can be scheduled to existing cluster or if a user needs more nodes. The methods must be usable by external components also. 

Adding this cmd under misc. "tools" with "estimate". At first, lets go with simple cpu, memory and count.

┆Issue is synchronized with this [Jira Task](https://datastax.jira.com/browse/K8CL-6) by [Unito](https://www.unito.io)
┆Issue Number: K8CL-6

### PR DESCRIPTION
**Pull request by burmanm**
Fixes #62

### CODE REVIEW

File: cmd/kubectl-k8ssandra/tools/estimate.go [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+}
+
+// Complete parses the arguments and necessary flags to options
+func (c *estimateOptions) Complete(cmd *cobra.Command, args []string) error {
+	return nil
+}
+
+// Validate ensures that all required arguments and flag values are provided
+func (c *estimateOptions) Validate() error {
+	if c.count < 0 {
```
Comment (adejanovski) [inline cmd/kubectl-k8ssandra/tools/estimate.go:101]: **Suggestion:** 0 should probably be considered invalid as well.
```suggestion
	if c.count <= 0 {
```

File: cmd/kubectl-k8ssandra/tools/estimate.go [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+		Example: fmt.Sprintf(estimateExample, "kubectl k8ssandra tools estimate"),
+		RunE: func(c *cobra.Command, args []string) error {
+			if err := o.Complete(c, args); err != nil {
+				return err
+			}
+			if err := o.Validate(); err != nil {
+				return err
+			}
+			if err := o.Run(); err != nil {
+				return err
```
Comment (adejanovski) [inline cmd/kubectl-k8ssandra/tools/estimate.go:66]: **Issue:** doing things this way always spits out the help text for the command in the console, along with the actual error:
```
k8ssandra-client % ./kubectl-k8ssandra tools estimate --count 25 --memory 2Gi --cpu 2000m 
Error: Unable to schedule the pods: unable to schedule all the pods, requested: 25, schedulable: 5
Usage:
  k8ssandra tools estimate [flags]

Examples:

        # Estimate if pods will fit the cluster
        kubectl k8ssandra tools estimate estimate [<args>]

        # Estimate if 4 pods each with 2Gi of memory and 2 vCPUs will be able to run on the cluster.
        # All CPU values and memory use Kubernetes notation
        kubectl k8ssandra tools estimate estimate --count 4 --memory 2Gi --cpu 2000m
        

Flags:
      --as string                      Username to impersonate for the operation. User could be a regular user or a service account in a namespace.
      --as-group stringArray           Group to impersonate for the operation, this flag can be repeated to specify multiple groups.
      --as-uid string                  UID to impersonate for the operation.
      --cache-dir string               Default cache directory (default "/Users/adejanovski/.kube/cache")
      --certificate-authority string   Path to a cert file for the certificate authority
      --client-certificate string      Path to a client certificate file for TLS
      --client-key string              Path to a client key file for TLS
      --cluster string                 The name of the kubeconfig cluster to use
      --context string                 The name of the kubeconfig context to use
      --count int                      new nodes to create
...
```

It makes it a little challenging to figure out what's going on.
Also we don't get a message if the pods can successfully be scheduled.

**Review by adejanovski — CHANGES_REQUESTED** [review]
Looks like something that will come in handy!
I have a few comments around the cli experience which could benefit some refinements.

File: cmd/kubectl-k8ssandra/tools/estimate.go (hunk 2 above)
Comment (burmanm) [reply cmd/kubectl-k8ssandra/tools/estimate.go]: Changed to this:

```console
➜  k8ssandra-client git:(add_scheduler_predictor) ✗ ./kubectl-k8ssandra tools estimate --count 6 --memory 2Gi --cpu 2000m    
2024/12/03 13:24:32 Pods can be scheduled to current cluster kind-kind
➜  k8ssandra-client git:(add_scheduler_predictor) ✗ ./kubectl-k8ssandra tools estimate --count 16 --memory 2Gi --cpu 2000m
2024/12/03 13:24:36 Error: Unable to schedule the pods to current cluster kind-kind: unable to schedule all the pods, requested: 16, schedulable: 6
```

**Pull request merged**